frost-secp256k1-tr/src/lib.rs

```diff
@@ -0,0 +1,628 @@
+
+    /// H1 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.1
+    fn H1(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        hash_to_scalar((CONTEXT_STRING.to_owned() + "rho").as_bytes(), m)
+    }
+
+    /// H2 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.2
+    fn H2(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        let mut hasher = tagged_hash("BIP0340/challenge");
+        hasher.update(m);
+        hasher_to_scalar(hasher)
+    }
+
+    /// H3 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.3
+    fn H3(m: &[u8]) -> <<Self::Group as Group>::Field as Field>::Scalar {
+        hash_to_scalar((CONTEXT_STRING.to_owned() + "nonce").as_bytes(), m)
+    }
+
+    /// H4 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.4
+    fn H4(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"msg", m])
+    }
+
+    /// H5 for FROST(secp256k1, SHA-256)
+    ///
+    /// [spec]: https://www.ietf.org/archive/id/draft-irtf-cfrg-frost-14.html#section-6.5-2.2.2.5
+    fn H5(m: &[u8]) -> Self::HashOutput {
+        hash_to_array(&[CONTEXT_STRING.as_bytes(), b"com", m])
+    }
+
+    /// HDKG for FROST(secp256k1, SHA-256)
+    fn HDKG(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "dkg").as_bytes(),
+            m,
+        ))
+    }
+
+    /// HID for FROST(secp256k1, SHA-256)
+    fn HID(m: &[u8]) -> Option<<<Self::Group as Group>::Field as Field>::Scalar> {
+        Some(hash_to_scalar(
+            (CONTEXT_STRING.to_owned() + "id").as_bytes(),
+            m,
+        ))
+    }
+
+    /// Generates the challenge as is required for Schnorr signatures.
+    fn challenge(R: &Element<S>, verifying_key: &VerifyingKey, msg: &[u8]) -> Challenge<S>
+    {
+        let mut preimage = vec![];
+        let tweaked_public_key = tweaked_public_key(&verifying_key.to_element(), &[]);
+        preimage.extend_from_slice(&R.to_affine().x());
+        preimage.extend_from_slice(&tweaked_public_key.to_affine().x());
+        preimage.extend_from_slice(msg);
+        Challenge::from_scalar(S::H2(&preimage[..]))
+    }
+
+    /// determine tweak is need
+    fn is_need_tweaking() -> bool {
+        true
+    }
+
+    /// aggregate tweak z
+    fn aggregate_tweak_z(
+        z: <<Self::Group as Group>::Field as Field>::Scalar,
+        challenge: &Challenge<S>,
+        verifying_key: &Element<S>,
+    ) -> <<Self::Group as Group>::Field as Field>::Scalar
+    {
+        let t = tweak(&verifying_key, &[]);
+        z + t * challenge.clone().to_scalar()
```

Comment:
I think at this point there's two different cases:

* `z + t * c`
* `z - t * c`

Reply:
so maybe this should be the code IIUC:

```rust
let tc = t * challenge.clone().to_scalar();
if verifying_key.y_is_odd() { z - tc } else { z + tc }
```